server/pulp/server/controllers/content.py

```diff
@@ -128,113 +137,392 @@ def refresh_content_source(content_source_id=None):
+            path_entry = report.data[UNIT_FILES][report.destination]
+            catalog_entry = path_entry[CATALOG_ENTRY]
+            self.validate_file(
+                catalog_entry.path,
+                catalog_entry.checksum_algorithm,
+                catalog_entry.checksum
+            )
+            path_entry[PATH_DOWNLOADED] = True
+            msg = _('{path} has already been downloaded.').format(
+                path=path_entry[CATALOG_ENTRY].path)
+            _logger.debug(msg)
+            raise SkipLocation()
+        except (InvalidChecksumType, VerificationException, IOError):
+            # It's either missing or incorrect, so download it
+            pass
+
+    def download_succeeded(self, report):
+        """
+        Marks the individual file for the unit as downloaded and moves it into
+        its final storage location if its checksum value matches the value in
+        the catalog entry (if present).
+
+        Inherited from DownloadEventListener.
+
+        :param report: the report associated with the download request.
+        :type  report: nectar.report.DownloadReport
+        """
+        content_unit = report.data[CONTENT_UNIT]
+        path_entry = report.data[UNIT_FILES][report.destination]
+
+        # Validate the file and update the progress.
+        catalog_entry = path_entry[CATALOG_ENTRY]
+        try:
+            self.validate_file(
+                report.destination,
+                catalog_entry.checksum_algorithm,
+                catalog_entry.checksum
+            )
+
+            relative_path = os.path.relpath(
+                catalog_entry.path,
+                FileStorage.get_path(content_unit)
+            )
+            if len(report.data[UNIT_FILES]) == 1:
+                # If the unit is single-file, update the storage path to point to the file
+                content_unit.set_storage_path(relative_path)
+                content_unit.import_content(report.destination)
+            else:
+                content_unit.import_content(report.destination, location=relative_path)
+            self.progress_successes += 1
+            path_entry[PATH_DOWNLOADED] = True
+        except (InvalidChecksumType, VerificationException, IOError), e:
+            _logger.debug(_('Download of {path} failed: {reason}.').format(
+                path=catalog_entry.path, reason=str(e)))
+            path_entry[PATH_DOWNLOADED] = False
+            self.progress_failures += 1
+        self.report_progress()
+
+        # Mark the entire unit as downloaded, if necessary.
+        download_flags = [entry[PATH_DOWNLOADED] for entry in
+                          report.data[UNIT_FILES].values()]
+        if all(download_flags):
+            _logger.debug(_('Marking content unit {type}:{id} as downloaded.').format(
+                type=content_unit.type_id, id=content_unit.id))
+            content_unit.downloaded = True
+            content_unit.save()
```

Comment:
This method uses several methods defined on the content unit. I see your point about race conditions, but it looks like under the hood, MongoEngine is doing the operation you suggested[0], and we're only modifying ``downloaded`` and ``last_modified`` on multi-file units. I can't think if a situation where this will be problematic (although that doesn't mean it doesn't exist, obviously).

One thing that could be done is to reload the content unit from the database in this method from the ``unit_id`` and ``type_id`` saved in the ``DownloadRequest.data`` dict (which was suggested above to avoid holding the entire list of content units in memory). 

[0] http://docs.mongoengine.org/apireference.html#mongoengine.Document.save